packages/grafana-ui/src/themes/stylesFactory.ts

```diff
@@ -0,0 +1,13 @@
+import memoizeOne from 'memoize-one';
+import { KeyValue } from '@grafana/data';
+
+/**
+ *  Creates memoized version of styles creator
+ * @param stylesCreator function accepting component's props and state and returning object with class names
+ */
+
+export function stylesFactory<TProps, TState, TResult extends KeyValue<string>>(
+  stylesCreator: (props: TProps, state?: TState) => TResult
```

Comment:
Not sure having props as argument is ideal here. The props will probably change a lot and will invalidate the memoization much more often even if you just need theme which almost never change. Not sure if we want to say exactly what has to be the type of the function, maybe we could just let it have any arguments the same way as memoizeOne has it's typing set up.

Reply:
@aocenas the type does not have to be the same as component's props. I.e. them is usually retrieved via context and is a single argument to majority of `getStyle` usages. So maybe the naming here is not precise. 

Maybe instead of `props, state` we could have `dependencies: TDeps` type which could be a  some kind of mixture of props/state if needed?

Reply:
1cb7a52

Reply:
Ah ok yeah that makes sense, the Props + State confused me here a bit :D